### ISSUE TITLE
Familiarisation with pythonic event loop in ROOT

### ISSUE THREAD

**Issue by mvesteri** [r-preston/MPhysProject2021#3]
@lmcairns @r-preston 

You can try the following python code
```python
#!/usr/bin/env python
from ROOT import gROOT
gROOT.SetBatch(True)
import ROOT

ch = ROOT.TChain('Z/DecayTree')
ch.Add('/storage/epp2/phshgg/DVTuples__v23/5TeV_2017_32_Down_EW.root')
```

Run with `python my_script.py`. A useful tip is that you can make the script "executable" by doing `chmod +x my_script.py`. Then you can just do `./my_script.py`. The `#!/usr...` line at the top of the script tells the operating system to execute as a python script.

Then try adding the following lines (one by one)
```
ch.Show() # list the branches in the `TTree`
ch.Show(0) # show the actual values of the first entry
ch.Scan() 
ch.Scan('mup_PT') # scan a specific branch
ch.Scan('mup_PT+mup_ETA') # scan a formula (though the one that I write is a bit silly)
```
Then try looping over the entries 
```python
for entry in ch:
     mup_PT = entry.mup_PT
     print(mup_PT)
```

The next step would be to try a more complicated calculation for each entry. E.g. the invariant mass of the mu+ and mu-

You can also try playing with python interactively with the `-i` option:
```bash
python -i my_script.py
```

**Comment by lmcairns-mphys** [comment]
![Screenshot 2020-10-15 130109](https://user-images.githubusercontent.com/72559837/96120527-8d404b00-0ee6-11eb-9a60-d2b33843f800.png)

**Comment by mvesteri** [comment]
Next exercise is to try and compute the Z mass in the python event loop.
```python
for entry in ch:
      mup_PT = entry.mup_PT # there is also ETA and PHI.
```

* ETA -> pseudorapidity
* PHI -> azimuthal angle

Suggest to first translate these to PX,PY,PZ. Also just use M = 0.105 GeV.
* for each muon we can write [PX,PY,PZ,E]. 
* ~`M = (P1 + P2)^2`. ~
* `M^2 = (P1 + P2)^2`. (note my correction 🤦 )

**Comment by mvesteri** [comment]
@lmcairns @r-preston just to add that I think you need to tag me in any post if you want me to get email notification.

**Comment by mvesteri** [comment]
@r-preston I saw that there was a question on the Z mass (email notifcation) but it is no longer here. Does that mean that you resolved it?

P.s. cc (@lmcairns)  regarding the invariant mass: this is not *that* different from vectors in Euclidian space. E.g. in Euclidian space you can translate/rotate a vector and it will still have the same "length". You can add them together and get a vector with a different length. The length is just the quadratic sum of the components. This might seem trivial (we learn pythagoras at at school etc..) but it is actually because of the Euclidian space. In relativity we work in a different (Minkowski) space. The (four)vectors now have four components rather than three. And the "length" (or invariant) rule is just slightly different. Instead of `sqrt(x**2 + y**2 + z**2)` we have `sqrt(E**2 -x**2 -y**2 -z**2)`. So the invariant mass is just the length of a vector in Minkowski space. If you think about it this way then you don't have to remember many formulae. You understand the derivation. The underlying maths is the same for Euclidian and Minkowski space. It is just that they have different "metrics", which is what changes the "length rule".

**Comment by r-preston** [comment]
> @r-preston I saw that there was a question on the Z mass (email notifcation) but it is no longer here. Does that mean that you resolved it?

Yes, it was a daft arithmetic mistake I'd overlooked. I was getting thrown off because the incorrect result was coincidentally about twice the expected one!

**Comment by mvesteri** [comment]
P.s. sorry I also realised that I wrote the wrong formula for two comments above

**Comment by mvesteri** [comment]
@r-preston OK now that you solved this I can tell you something annoying 😄 .
There is a cheat way with ROOT's `TLorentzVector`.
```python
mup = ROOT.TLorentzVector(mup_PX,mup_PY,mup_PZ,mup_E)
mum = ...
mass = (mup+mum).M()
# and there are lots of other helpful functions...
```

**Comment by mvesteri** [comment]
You can also directly use the `PT,ETA,PHI` variables like this
```python
muon_mass = 0.105 # GeV
mup = ROOT.TLorentzVector()
mup.SetPtEtaPhiM(mup_PT*1.e-3,mup_ETA,mup_PHI,muon_mass)
```

P.s. while the original `TTree` files have energy/momenta in units of MeV I would recommend translating as early as possible to GeV, which is a more natural unit for W and Z decays. (LHCb likes MeV which is more natural for its lower energy beauty/charm hadron program)

**Comment by lmcairns-mphys** [comment]
@mvesteri Managed to work out the formula yesterday evening and wrote some code in python, just had some issues working out how to plot histograms in ROOT but I've solved that now. I also had a go at using the TLorentzVector yesterday, and although I don't think my code is the most efficient seeing your latest comment (I'll have another go at some point), it gets the same results as my formula.

I've attached the histograms and photos of my invariant mass calculation (I'll take better ones for the weekly record).
I will post my weekly record at some point over the weekend with my workings and notes up to this point, and the full codes I wrote for these.

[InvariantMassCalc.pdf](https://github.com/r-preston/MPhysProject2021/files/5392398/InvariantMassCalc.pdf)
[ZMassPlot_Maths.pdf](https://github.com/r-preston/MPhysProject2021/files/5392340/ZMassPlot_Maths.pdf)
[ZMassPlots_ROOT.pdf](https://github.com/r-preston/MPhysProject2021/files/5392342/ZMassPlots_ROOT.pdf)
